api/client/build.go

```diff
@@ -427,20 +488,64 @@ func (cli *DockerCli) makeTrustedDockerfileFrom(input io.Reader, approximateSize
 			}
 			ref := registry.ParseReference(tag)
 
-			if !ref.HasDigest() {
+			if !ref.HasDigest() && isTrusted() {
 				trustedRef, err := cli.trustedReference(repo, ref)
 				if err != nil {
-					return nil, err
+					return nil, fmt.Errorf("unable to resolve image (%q) to trusted digest reference: %s", ref.ImageName(repo), err)
 				}
 
 				line = dockerfileFromLinePattern.ReplaceAllLiteralString(line, fmt.Sprintf("FROM %s", trustedRef.ImageName(repo)))
 			}
 		}
 
-		if _, err := fmt.Fprintln(dockerfileBuf, line); err != nil {
+		if _, err := fmt.Fprintln(buf, line); err != nil {
 			return nil, err
 		}
 	}
 
-	return dockerfileBuf, scanner.Err()
+	return buf, scanner.Err()
+}
+
+func trustedDockerfileTarWrapper(inputTarStream io.Reader, newDockerfile *bytes.Buffer, dockerfileName string) io.Reader {
+	pipeReader, pipeWriter := io.Pipe()
+
+	tarReader := tar.NewReader(inputTarStream)
+
+	go func() {
+		tarWriter := tar.NewWriter(pipeWriter)
+
+		for {
+			hdr, err := tarReader.Next()
+			if err == io.EOF {
+				// Signals end of archive.
+				tarWriter.Close()
+				pipeWriter.Close()
+				return
+			}
+			if err != nil {
+				pipeWriter.CloseWithError(err)
+				return
+			}
+
+			var content io.Reader = tarReader
+
+			if hdr.Name == dockerfileName {
```

Comment:
Is the name a strong enough check? I think it is but this might be worth a note here that indicates it is and why.